QuizApp/NetworkService.swift

```diff
@@ -0,0 +1,133 @@
+import Foundation
+import Reachability
+
+
+class NetworkService : NetworkServiceProtocol {
+    let baseUrl = "https://iosquiz.herokuapp.com/api"
+    
+    func executeUrlRequest<T: Decodable>(_ request: URLRequest, completionHandler:
+                                            @escaping (Result<T, RequestError>) -> Void) {
+        let dataTask = URLSession.shared.dataTask(with: request) { data, response,
+                                                                   err in
+            
+            guard err == nil else {
+                completionHandler(.failure(.clientError))
+                return
+            }
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                  (200...299).contains(httpResponse.statusCode) else {
+                completionHandler(.failure(.serverError))
+                return
+            }
+            guard let data = data else {
+                completionHandler(.failure(.noDataError))
+                return
+            }
+            guard let value = try? JSONDecoder().decode(T.self, from: data) else {
+                completionHandler(.failure(.decodingError))
+                return
+            }
+            completionHandler(.success(value))
+        }
+        dataTask.resume()
+    }
+    
+    func login(email: String, password: String, onCompletion: @escaping (LoginStatus)->Void) {
+        let reach = Reachability.init(hostName: baseUrl)
+        if !(reach?.isReachable() ?? false) {
+            onCompletion(.error(1, "No network"))
+            return
+        }
+        
+        var url = URLComponents(string: "\(baseUrl)/session")!
```

Comment:
This should also be in a `guard` statement. I know you're sure the URL is valid, but force unwraps are highly disencuraged.